pkg/provisioner/controller.go

```diff
@@ -425,14 +399,33 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 	} else {
 		ob, err = c.provisioner.Grant(options)
 	}
+	// Record whether the provisioner returned an empty object bucket for error handling use later
+	emptyBucket := (ob == nil || ob == (&v1alpha1.ObjectBucket{}))
+
+	// Set information about the OB as soon as possible, even before handling errors so that the
+	// deferred cleanup function is not missing critical information needed to clean up.
+	// More note on above: can't check if provisioner returned empty OB after we set OB info below.
+	// The returned OB should only be nil if there is also an error with provisioning; in such a
+	// case, the provisioner should have cleaned up after itself already so we do not need to set
+	// info for OB cleanup.
+	if ob != nil {
+		setObjectBucketName(ob, key)
+		ob.Spec.StorageClassName = obc.Spec.StorageClassName
+		if ob.Spec.ReclaimPolicy == nil || *ob.Spec.ReclaimPolicy == corev1.PersistentVolumeReclaimPolicy("") {
+			// Do not blindly overwrite the reclaim policy. The provisioner might have reason to
+			// specify a reclaim policy that is  different from the storage class.
+			ob.Spec.ReclaimPolicy = options.ReclaimPolicy
+		}
+		ob.SetLabels(c.provisionerLabels)
+	}
+
 	if err != nil {
 		return fmt.Errorf("error %s bucket: %v", verb, err)
-	} else if ob == (&v1alpha1.ObjectBucket{}) {
+	} else if emptyBucket {
```

Comment:
of course, if the ob != nil then we've updated it above so it's no longer empty but I get the logic...